### PR TITLE
[PHI decouple] move dropout_impl and cuda_graph_with_memory_pool from fluid to phi

### PR DESCRIPTION
**Pull request by huangjiyi**
### PR types

Others

### PR changes

Others

### Describe

- #47615 
- move `dropout_impl.cu.h` and `dropout_impl_util.h` from fluid to phi.
- move part of `cuda_graph_with_memory_pool.h` from fluid to phi.
- move whole `cuda_graph.h` from fluid to phi.

### CODE REVIEW

**Comment by huangjiyi** [comment]
@YuanRisheng 帮忙 Review 一下

File: paddle/phi/kernels/funcs/dropout_impl_util.h [hunk 1]
```diff
@@ -27,13 +29,11 @@ inline void GetSeedDataAndIncrement(const phi::GPUContext& dev_ctx,
...
                                     uint64_t* increment) {
-  int device_id = dev_ctx.GetPlace().GetDeviceId();
-  auto gen_cuda = framework::DefaultCUDAGenerator(device_id);
+  auto gen_cuda = dev_ctx.GetGenerator();
 
   if (seed) {
     phi::DenseTensor seed_cpu_tensor;
-    paddle::framework::TensorCopySync(
-        *seed, platform::CPUPlace(), &seed_cpu_tensor);
+    phi::Copy(dev_ctx, *seed, phi::CPUPlace(), false, &seed_cpu_tensor);
```
Comment (YuanRisheng) [inline paddle/phi/kernels/funcs/dropout_impl_util.h]: 这里是同步拷贝，对应phi下copy参数应该是true

Comment (huangjiyi) [reply paddle/phi/kernels/funcs/dropout_impl_util.h]: 已处理

**Comment by huangjiyi** [comment]
@luotao1 帮忙看一下 CI-Mac-Python3 的问题，重跑了几次都一样。

**Comment by luotao1** [comment]
@YuanRisheng 可以review了

**Review by YuanRisheng — APPROVED** [review]

**Pull request merged**